### PR TITLE
Don't emit structure padding members if no padding is required.

### PR DESCRIPTION
**Pull request by DiamondLovesYou**
Previously, we would include an array in between every structure member + another before the first member. The vast majority of the time these array types would be zero sized, and would serve no purpose to LLVM, as we already ensure proper member alignment.

This change *might* be faster, due to fewer structure members, but I didn't bench.

(Zero length arrays are disallowed in SPIR-V. Avoiding them here is slightly easier than trying to patch things up on the LLVM side.)

### CODE REVIEW

**Review by ecstatic-morse — COMMENTED** [review]
Overall this looks quite nice to me. It's not ideal to have *two* mappings from source-field order to `u32` with `memory_index` and now `padded_indices`. Preferably we would use a newtyped index for these. However, I don't think this needs to block this PR as it was preexisting.

I've been trying to ease their review burden, but since this is pretty squarely in their domain, cc @eddyb.

**Comment by ecstatic-morse** [comment]
Sweet! This seems to be a small win across the board, and (very slightly) faster `check` builds indicate that `rustc` itself. I'm not concerned with the `deep-vector-opt` regression since that benchmark does basically [no codegen](https://github.com/rust-lang/rustc-perf/blob/master/collector/benchmarks/deep-vector/src/main.rs), although if you have a theory as to why it regressed I'm happy to hear it.

File: src/librustc_target/abi/mod.rs [hunk 1]
```diff
@@ -656,6 +656,11 @@ pub enum FieldsShape {
         // FIXME(eddyb) build a better abstraction for permutations, if possible.
         // FIXME(camlorn) also consider small vector  optimization here.
         memory_index: Vec<u32>,
+
+        /// Map source order indices to memory indices which include
+        /// required implicit padding, eg due to alignment.
+        /// If None, then this is the identity.
+        padded_indices: Option<Vec<u32>>,
```
[the remark this answers is not in the document]
Comment (DiamondLovesYou) [reply src/librustc_target/abi/mod.rs:663]: I considered something like that, but I felt that a proper fix would be creating a proper `Bijection` collection/map in librustc_data_structures (which I thought a bit about doing, but I didn't want to that refactor in *this* PR, sooo), and I didn't want to introduce a new structure (`Permutation` as you say) which would just get dumpstered Shortly(TM).

I'll admit maybe I should have not used `padded_indices` directly elsewhere (`struct_llfields`).. I've fixed locally (and made it a bit more efficient by storing the count), waiting on the tests.

> We could also use this for memory_index above in a follow-up PR to resolve the FIXME and save a few heap allocations.

Actually, I would say a higher priority should be a proper bijection data structure. I've given this some thought and I think it's totally possible to support iterating both directions (forward and inverse) with just a single Vec, but I'm not sure I want to set aside more of my free time to implement it and fix librustc_middle::ty::layout to then use it (life of a hobbyist :frowning_face:). Nevertheless, with such a structure, `memory_index` would just be the permutation of the bijection map, and thus could be deleted in favor of the map.

**Comment by DiamondLovesYou** [comment]
Cool! :smile: 

Not sure about `deep-vector-opt`.. It's not like there's a lot of LLVM types/layouts to be created (at least not directly by the test), so prima facie this patch shouldn't boost speeds on that test anyway. Odd that it would regress (maybe the win is only after a certain crate size threshold?).

I've got one last slight optimization (tests..), then this should be ready. Now the padded field count (really, the upper bound) is precomputed along with the padded indices themselves.

File: src/librustc_target/abi/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ecstatic-morse) [reply src/librustc_target/abi/mod.rs]: In that case, r? @eddyb. I feel strongly that `Option::None` should not have special semantics, but I also don't own the code.

**Comment by eddyb** [comment]
> The vast majority of the time these array types would be zero sized, and would serve no purpose to LLVM, as we already ensure proper member alignment.

They do serve a purpose though, and that is having simple to calculate field indices, without needing a layout-keyed cache. I guess from the other comments (and, well, this PR working at all) that this is taken into account (although `rustc_target` shouldn't encode LLVM-specific details IMO).

I'd much rather stop using LLVM's "struct" types and just do byte offsets instead of regular GEPs (but SPIR-V will probably like that even less).

I wish the LLVM devs would understand that there is nothing good about their "types" (no, debugging LLVM IR shouldn't be done semantically) but I don't have the time to constantly try to convince them, and even the things they agree on (pointee types) have barely made any progress in the past 5 years.

<hr/>

I don't think supporting SPIR-V is a plausible goal for the Rust LLVM backend, given all the limitations I've heard about. It would require a lot of transformation of anything involving memory IIRC.

cc @MaikKlein because of their experience building [`rlsl`](https://github.com/MaikKlein/rlsl)

cc @rust-lang/compiler

**Comment by nikomatsakis** [comment]
Repeating what I said in today's compiler meeting:

I would love to see better support for targeting GPUs with Rust, and SPIR-V sounds like a plausible candidate, but I think it'd be good to have a better view of the overall plan before we make changes that seem to complicate the compiler (and don't yield other wins). It seems quite plausible that we'll get stuck with a more complex compiler and without real SPIR-V support otherwise. An MCP (or even ultimately RFC, depending on the scope) seems appropriate.

**Comment by pnkfelix** [comment]
[discussed at yesterday's T-compiler meeting](https://zulip-archive.rust-lang.org/238009tcompilermeetings/55560weeklymeeting2020060454818.html#199752860)

As @nikomatsakis wrote above, we really would need a separate proposal describing the overall plan for adding such support. We don't want to proceed by making piecemeal changes that don't get us all the way that don't have other motivations beyond such hypothetical support.

So, I'm closing this PR, with the hope that we can convince people enthused about a GPU backend to explain in an MCP or RFC how we are going to get there.

**Pull request closed without merging**